### PR TITLE
PIONEER DDJ-SX mapping bugfixes

### PR DESCRIPTION
**Pull request by DJMaxergy**
Here are some fixes to bugs (and optimizations) I recognized over the last few months.

### CODE REVIEW

**Comment by Be-ing** [comment]
Could you rebase this on the 2.1 branch so we can include it in 2.1.1?

**Comment by DJMaxergy** [comment]
Ok...it seems that I'm too stupid for that:
Directly rebasing to origin/2.1 fails because of many files which have to be merged somehow.
What commands do I have to execute to rebase it correctly?

**Comment by uklotzde** [comment]
Git offers a range-based rebase and cherry-pick commands, but these are too complicated for a few commits:
1) Rename your current branch, e.g. 'pioneerDDJSXmapping_bugfixes_bak'
2) Create a new local branch from 'upstream/2.1' named 'pioneerDDJSXmapping_bugfixes'
3) Cherry-pick your commits from the backup branch one by one
4) Force push the rebased branch into your GitHub repo

**Comment by uklotzde** [comment]
By re-using the original name this PR is preserved. Afterwards just change the merge target and you should only see your 5 commit here again.

**Comment by Be-ing** [comment]
> Cherry-pick your commits from the backup branch one by one

It is easier to do `git rebase -i upstream/2.1` for this step and remove commits that are not related to this branch.

**Comment by uklotzde** [comment]
You did it :+1:

LGTM. I propose to merge this, even though we don't have feedback from any other owner.

Does the Wiki page need to be updated?

**Comment by DJMaxergy** [comment]
?? I didn't do anything yet. I now try this cherry-pick thing. I hope, I don't do a mess...

**Comment by DJMaxergy** [comment]
> Create a new local branch from 'upstream/2.1' named 'pioneerDDJSXmapping_bugfixes'

How can I do that (using which exact command)? Sorry, but I don't get it at the moment.

**Comment by DJMaxergy** [comment]
Ok...I did the following: 
`git checkout -b pioneerDDJSXmapping_bugfixes upstream/2.1`

But in this branch there aren't any Pioneer DDJ-SX mapping files. That's why I cannot cherry-pick the changes I recently made. Why are there no files for the mapping?

**Comment by DJMaxergy** [comment]
It seems that it's an old state from March 2017 where there didn't exist the mapping. I'm totally confused now.

**Comment by uklotzde** [comment]
If you are not familiar with Git cherry-picking might be easier than an interactive rebase.

After creating and checking out your new branch (as you already did) you can cherry-pick visually from your existing branch by using the *git gui*:  *Repository* > *Visualize All Branch History*, select a commit, *Right click on commit message* > *Cherry-pick this commit*

Or use *git cherry-pick <commit >* on the command line.

Repeat for each of your 5 commits.

**Comment by DJMaxergy** [comment]
> After creating and checking out your new branch (as you already did) you can cherry-pick visually from your existing branch by using the git gui: Repository > Visualize All Branch History, select a commit, Right click on commit message > Cherry-pick this commit

Or use git cherry-pick on the command line.

When I do that, Git wants to commit the whole mapping file because it detects that the file doesn't exist. Is this correct then?

**Comment by uklotzde** [comment]
*git checkout -b* has reset your local branch! Now you have to cherry pick the commits from your remote branch *origin/pioneerDDJSXmapping_bugfixes*. Don't push your local branch until you are done!

That's why I recommended to rename the original branch as a backup first ;)

**Comment by DJMaxergy** [comment]
> That's why I recommended to rename the original branch as a backup first ;)

I did that!
I'm just wondering why there isn't any existing mapping file in the new local branch (based on upstream/2.1). The mapping was added in May 2017.
In the master branch the mapping files already exist.

Is it correct that the mapping files are missing in the new local branch based on upstream/2.1?

**Comment by uklotzde** [comment]
No, the mapping is present in upstream/2.1.

Try the rebasing as an exercise. Otherwise you can find your rebased commits here:
https://github.com/uklotzde/mixxx/tree/pioneerDDJSXmapping_bugfixes

**Comment by Be-ing** [comment]
I just rewrote the [Using Git wiki page](https://mixxx.org/wiki/doku.php/using_git) to explain the typical workflow more succinctly.

**Comment by DJMaxergy** [comment]
It doesn't matter what I do....if I pull upstream/2.1 I don't have the mapping files present. So if I do the cherry-pick of my bug fixes Git wants to commit the whole mapping file.
I also tried:
`git cat-file -e upstream/2.1:res/controllers/Pioneer-DDJ-SX-scripts`
Result:
`fatal: Not a valid object name upstream/2.1:res/controllers/Pioneer-DDJ-SX-scripts`
Other existing mapping files can be found with this command.

I really don't know what I can do here now. I don't want to mess everything up.

**Comment by Be-ing** [comment]
I don't know what you have configured as your `upstream` remote. Follow the directions [here](https://mixxx.org/wiki/doku.php/using_git#set_up_git) to properly configure the `upstream` remote.

**Comment by DJMaxergy** [comment]
> I don't know what you have configured as your upstream remote. Follow the directions here to properly configure the upstream remote.

Yes, I've just read your info. I will go through the description.

**Comment by DJMaxergy** [comment]
Thank you for all the help! Sorry that I'm totally not familiar with Git.

**Comment by DJMaxergy** [comment]
Thank you @Be-ing for rewriting the [wiki page](https://mixxx.org/wiki/doku.php/using_git). That helped me a lot!
It's nearly done...I have one last problem when I want to push the branch into my GitHub repo:
`To https://github.com/DJMaxergy/mixxx.git
 ! [rejected]        pioneerDDJSXmapping_bugfixes -> pioneerDDJSXmapping_bugfixes (non-fast-forward)
error: failed to push some refs to 'https://github.com/DJMaxergy/mixxx.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.`

What do I have to do here?

**Comment by uklotzde** [comment]
You need to do a *force push* after rebasing to advice Git to replace the existing commits in the remote branch.

**Comment by uklotzde** [comment]
:+1: I just changed the target from *master* to *2.1*

File: res/controllers/Pioneer-DDJ-SX-scripts.js [hunk 1]
```diff
@@ -97,8 +97,8 @@ PioneerDDJSX.samplerCueGotoAndPlay = false;
 PioneerDDJSX.autoPFL = true;
 
 // If true, new in Mixxx 2.1 introduced library controls will be used,
-// else old playlist controls will be used (default: false).
-PioneerDDJSX.useNewLibraryControls = false;
+// else old playlist controls will be used (default: true).
+PioneerDDJSX.useNewLibraryControls = true;
```
Comment (Be-ing) [inline res/controllers/Pioneer-DDJ-SX-scripts.js]: Is this option necessary? Can we remove it?

Comment (Be-ing) [reply res/controllers/Pioneer-DDJ-SX-scripts.js]: Considering you just changed the shifted behavior to act like the unshifted behavior when this option is false, I think the option should be removed.

Comment (DJMaxergy) [reply res/controllers/Pioneer-DDJ-SX-scripts.js]: You're right! I removed the option and its If-else dependencies in the library concerned functions.

**Comment by Be-ing** [comment]
LGTM. Thank you for maintaining this.

**Pull request merged**